src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php

```diff
@@ -289,6 +290,10 @@ public function addCartRuleAction(Request $request, int $cartId): JsonResponse
     {
         $cartRuleId = $request->request->getInt('cartRuleId');
         try {
+            $cartInfo = $this->getCartInfo($cartId)->getShipping();
```

Comment:
It's weird to block the cart rule addition if the order has free shipping It means as long as you don't pay the shipping you can't add have additional discounts on products?

Are you sure this code shouldn't be in the `setFreeShippingAction` instead? In any case it would be better to manage this check in a handler rather than in the controller, the domain rules should be implemented by the handler It makes behat testing easier, and if the handler is later used in an API it contains the whole domain assertions

Reply:
You just spotted a bug I've left: It means as long as you don't pay the shipping you can't add have additional discounts on products! :man_astronaut:

Reply:
Just I think the corrected code should go to `AddCartRuleToCartHandler` because that handler is actually invoked when adding any cart rule as a voucher